src/mlpack/methods/ann/layer/nearest_interpolation_impl.hpp

```diff
@@ -0,0 +1,228 @@
+#define MLPACK_METHODS_ANN_LAYER_NEAREST_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "nearest_interpolation.hpp"
+
+namespace mlpack {
+
+template<typename MatType>
+NearestInterpolationType<MatType>::NearestInterpolationType():
+  Layer<MatType>(),
+  inRowSize(0),
+  inColSize(0),
+  outRowSize(0),
+  outColSize(0),
+  depth(0),
+  batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>::
+NearestInterpolationType(const size_t inRowSize,
+                         const size_t inColSize,
+                         const size_t outRowSize,
+                         const size_t outColSize,
+                         const size_t depth) :
+  Layer<MatType>(),
+  inRowSize(inRowSize),
+  inColSize(inColSize),
+  outRowSize(outRowSize),
+  outColSize(outColSize),
+  depth(depth),
+  batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>::
+NearestInterpolationType(const NearestInterpolationType& other) :
+  Layer<MatType>(),
+  inRowSize(other.inRowSize),
+  inColSize(other.inColSize),
+  outRowSize(other.outRowSize),
+  outColSize(other.outColSize),
+  depth(other.depth),
+  batchSize(other.batchSize)
+{
+  // Nothing to do here. 
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>::
+NearestInterpolationType(NearestInterpolationType&& other) :
+  Layer<MatType>(std::move(other)),
+  inRowSize(std::move(other.inRowSize)),
+  inColSize(std::move(other.inColSize)),
+  outRowSize(std::move(other.outRowSize)),
+  outColSize(std::move(other.outColSize)),
+  depth(std::move(other.depth)),
+  batchSize(std::move(other.batchSize))
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>&
+NearestInterpolationType<MatType>::
+operator=(const NearestInterpolationType& other) 
+{
+  if (&other != this) {
+    Layer<MatType>::operator=(other);
+    inRowSize = other.inRowSize;
+    inColSize = other.inColSize;
+    outRowSize = other.outRowSize;
+    outColSize = other.outColSize;
+    depth = other.depth;
+    batchSize = other.batchSize;
+  }
+  return *this;
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>&
+NearestInterpolationType<MatType>::
+operator=(NearestInterpolationType&& other) 
+{
+  if (&other != this) {
+    Layer<MatType>::operator=(std::move(other));
+    inRowSize = std::move(other.inRowSize);
+    inColSize = std::move(other.inColSize);
+    outRowSize = std::move(other.outRowSize);
+    outColSize = std::move(other.outColSize);
+    depth = std::move(other.depth);
+    batchSize = std::move(other.batchSize);
+  }
+  return *this;
+}
+
+template<typename MatType>
+void NearestInterpolationType<MatType>::Forward(
+  const MatType& input, MatType& output)
+{
+  batchSize = input.n_cols;
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, batchSize);
```

Comment:
I think that this should not be necessary---`output` should already be sized correctly.  So you can remove this.